### ISSUE TITLE
Segfault doing trip call

### ISSUE THREAD

**Issue by Kentaree** [Project-OSRM/osrm-backend#1747]
When doing a trip call with a large amount of points (760 in this case) on version v4.8.1
I consistently get a segfault with coredump. I recompiled with debug, and this is the error that it's getting:

```
**** Internal Program Error - assertion (trip_dist >= 0) failed in std::pair<int, __gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int> > > osrm::trip::GetShortestRoundTrip(NodeID, const DistTableWrapper<int>&, std::size_t, std::vector<unsigned int>&):
/opt/osrm/library/../plugins/../algorithms/trip_farthest_insertion.hpp(79): previous trip was not minimal. something's wrong
```

**Comment by daniel-j-h** [comment]
For the record: [trip_farthest_insertion.hpp](https://github.com/Project-OSRM/osrm-backend/blob/3f82570469719d8350320a5e23b342945dc57bba/algorithms/trip_farthest_insertion.hpp#L80).

/cc @chaupow

**Comment by Kentaree** [comment]
To provide more info, I'm using the Europe maps, and I've managed to reproduce it on a smaller batch:

```
/trip?alt=false&geometry=false&loc=54.825438,-1.705953&loc=54.825552,-1.705978&loc=54.825585,-1.706648&loc=54.825892,-1.705887&loc=54.825597,-1.706685&loc=54.82650010295488,-1.7056939477946602&loc=54.826137,-1.707585&loc=54.826875,-1.705677&loc=54.827075,-1.705597&loc=54.827548,-1.70535&loc=54.827563,-1.705365&loc=54.826189409129086,-1.7110350269414407&loc=54.826832,-1.711137&loc=54.826842,-1.711217&loc=54.826847,-1.711207&loc=54.828115,-1.709803&loc=54.827847,-1.71037&loc=54.828123,-1.709932&loc=54.828475,-1.709077&loc=54.82748,-1.71097&loc=54.827488,-1.710953&loc=54.826297,-1.711922&loc=54.82823,-1.709763&loc=54.827502,-1.710975&loc=54.828402,-1.709502
```

**Comment by chaupow** [comment]
Thanks Pieter! I'll look into it asap.

**Comment by chaupow** [comment]
Further breaking down the problem:
```
http://localhost:5000/trip?alt=false&geometry=false&loc=54.827548,-1.70535&loc=54.827563,-1.705365&loc=54.828123,-1.709932&loc=54.828475,-1.709077&loc=54.82748,-1.71097&loc=54.827488,-1.710953&loc=54.826297,-1.711922&loc=54.82823,-1.709763&loc=54.827502,-1.710975&loc=54.828402,-1.709502
```
leads to the same segfault.

_______________

The distance table is

![image](https://cloud.githubusercontent.com/assets/804027/10721028/dc401490-7be0-11e5-9ac4-2db31725e658.png)
______________
One of the problems is:
distance `9 -> 0 -> 1` is `1356 + 0` and
distance `9 -> 1` is `1357` 
but a distance with an additional via node should never be faster than the distance without a via node (@TheMarex got any clue?)
______________
@Kentaree. there's a quick dirty fix here but it does not fix the real problem:
https://github.com/Project-OSRM/osrm-backend/tree/fix/trip_segfault

I'll now look into the strange distance table.

**Comment by TheMarex** [comment]
@chaupow I think this is a rounding problem.

When a coordinate snaps to a segment, the duration is adjusted to match the projected position of the coordinate.

https://github.com/Project-OSRM/osrm-backend/blob/bb49e03435c4f71afbfc2bbbc645af8a8c168bed/data_structures/static_rtree.hpp#L1098

For example: Lets say you have an edge `(u, v)` with duration 100. If you place a coordinate exactly in the middle of the segment yielding `(u, v')`, the adjusted duration will be `100 * 0.5 = 50`. Now imagine two coordinates. One placed at 0.99 and one at 0.999. This means `(u, v')` now has a duration of `100 * 0.99 = 99`, but `(u, v'')` also has a duration of `100 * 0.995 = 99`. In which case `(v', v'')` has a duration of 0.

**Comment by KieCap** [comment]
Hi guys,
Any update on this issue yet? Has a fix been added to master or develop yet?
Thanks,
Kieran

**Comment by TheMarex** [comment]
@KieCap addressed in 94651744af861473a956dd611325ce840f6c668b

**Comment by KieCap** [comment]
Thanks @TheMarex
